### ISSUE TITLE
[Idea] - Restoring and saving user defined sessions

### ISSUE THREAD

**Issue by ldelossa** [luakit/luakit#946]
**Current Behavior:**
Using the session module, I cannot seem to save a user-defined session and load it via a user-defined function in userconf.lua

**Desired Behavior:**
I'd like to use the session api to save a session file, in my case "work". Then write a user-defined function in userconf.lua which will restore the user-defined session.

I do not want any other tabs which maybe open "sticking" in this session, so `:wq` is not great for my use case.

**How can we reproduce it (step by step):**

I create a number of tabs I'd like to keep as a user-defined "work" session. 

I then utilize the "session.save" api to store this file at "~/.config/luakit/session/work"

I then wrote this function in my `userconf.lua` file

```
-- userconf.lua

require "vertical_tabs"

function work_sesh ()
    require("session").load(false, "/home/louis/.config/luakit/sessions/work")
    require("session").restore(false)
end
```

Unfortunately running this function via `:lua work_sesh()` did not produce the desired effect of opening a `luakit` session with my user-defined work session.

Is this a supported use-case or am I going about this all wrong?

**Environment:**

Linux Distribution & Version: Fedora 35
Output of `luakit --version`:

```
luakit 2.3
  built with webkit 2.32.3 (installed version: 2.32.3)
```

**Note about webkit issues:**

If you're reporting a rendering issue, please test it with the gnome
browser ephiphany as well. If the issue occurs there too, we're very
likely not able to help. These issues should be reported to webkit:
https://bugs.webkit.org

**Comment by ldelossa** [comment]
Got this to work in kinda a hacky way: 

```
function work()
    tmp = session.session_file
    session.session_file = "/home/louis/.config/luakit/sessions/work"
    session.restore(false)
    session.session_file = tmp
end
```

**Comment by MuhammedZakir** [comment]
`session.load` does NOT restore session, it just returns a table containing your session.
> `session.load (delete, file)`
> 
> Load session state from a file, and optionally delete it.
> 
> The session state is **not restored**. This function only **loads the state into a table and returns it**.

`session.resore` uses `session.session_file` to restore the saved session, and if it fails, then uses `session.recovery_file`. That's why, the method in your comment worked. 

---

Currently, AFAIK, you can't load a session from a table, neither can you restore from a particular file. I am new to LuaKit, so I don't know if there was a reason for doing it this way.

For now, simplest way to get what you want is to export `restore_file` function in `libs/session.lua`. Change
```
local restore_file = function (file, delete)
```

to
```
_M.restore_file = function (file, delete)
```

**Comment by MuhammedZakir** [comment]
IMO, `restore` function\* should be changed to take an optional second argument, either a `string` (filename), or a table (`session.load` result). Or define seperate functions?

\* `restore_file` function will have to be changed too.

**Comment by ldelossa** [comment]
Yes i agree, if "user-defined" sessions is something LuaKit is interested in, i wouldn't mind hacking this out as a new additional feature in the session module.

My goal here is accomplished for now, but it could be a first class citizen with a well known session directory sessions get stored too, and a "restore_user_session" function that just takes a name. Simple stuff, just makes it easier. 

Ill wait for a bit of feedback, but id like to do this.

**Comment by taobert** [comment]
Can you give more information about how you want it to work?
From what you've described, (and the fact that you don't have a `session.save(luakit.data_dir .. "/not-work")` in your function,) i'm imagining that you'll load your work tabs, gradually pollute that space with other (research?) tabs,
and then periodically want to blow it all away and revert to the base set of work tabs.
Looking at your function, `window.` is poorly formed, so i guess something got cut off. (`window.close_win`?)


If you're after a minimum-complexity solution, your "hacky" function is probably pretty good. 
As the session loading is a large part of the startup time, alternatives would include:
- restart without saving the session and have your function call `new_tab` to load what you want
- restart and have your `userconf.lua` override the session file, based on some specified criteria, to load the work session.

You may also find that the `tabgroups` plugin gives you behaviour close enough to what you're after. (I use it to segregate stuff by topic.)


The current session handling is designed to save the state over a restart,
so doing anything other than that with it will either have to work around what it currently does,
or require some reworking of how that infrastructure operates.
I think this is a great thing for people to be thinking about: Having more flexible and filigreed session handling feels like the sort of thing that would be useful for a number applications. (I've often thought that it'd be nicer/faster/smaller, to be able to load only the tabgroup(s) i'm working with at the moment, and be able to unload tabgroups that i'm unlikely to use for a while.)
I'm unconvinced that deep changes to `session.lua` are the best solution to your workflow as i understand it,
but jettisoning and reloading entire sessions is a fairly blunt instrument.
Reusing tabs (webviews) which are already loaded would probably be faster.




To reiterate @MuhammedZakir's reading of the code,
`restore` just calls `restore_file` with a filename, which calls `load` to get the table of windows and tabs to set up.
Once upon a time these were all available externally to the module.
It appears that when the session module got rejigged to the now-preferred `_M` format (`103b1f4bf`) `restore_file` was made `local`. I don't see a clear reason why it _should_ be local (none of the other `session` functions are) so my supposition is that it was an oversight.
As @MuhammedZakir points out, if `local restore_file` were changed to `_M.restore_file`,
then the session file juggling in the "hacky" function could be removed, and `restore_file` called directly.
I think this would be a defensible PR.

**Comment by ldelossa** [comment]
@taobert looking over the code, what I'm thinking is a rough outline (take all function naming as temporary)

- new exported method is added to session module `save_user_session(name)` 
  - this method will simply wrap the current session.save method and point the file to `$configdir/sessions/$name`
- a new exported method is added named `load_user_session(name)` which can just internally call the `restore_file` function with the path `$configdir/sessions/$name`

That's all I'm thinking, and it seems pretty simple to me. There needs to be some extra stuff like listing and deleting sessions etc... but I think thats self explanitory for now.

My goal is to actually not touch anything around how sessions work today, just piggy back off the state management that's going on.

p.s : also fixed function above to what I'm using today.

TBH I'm too new to understand how this: 
> Reusing tabs (webviews) which are already loaded would probably be faster.

and tab groups work. 

If you feel like the above suggestions are "better" or can be imagined in the same way as hacking on the session module I'll take your suggestion.

**Comment by taobert** [comment]
Currently a session file can store multiple tabs for multiple windows.
When `session.save` is called, it makes a session file with all tabs for all windows you have open.
Every time you load a session (with `restore_file`) each window in the session file gets opened with all of it's associated tabs.
There's currently no capability to remove the old state, because it's only ever called at startup.
So, currently, saving and then loading a session will duplicate everything you have open.
This is probably not what you want.



The current `session.save` takes a filename, as does `restore_file` (which i feel should be corrected to be in `_M` rather than `local`). So adding new functions to `session` which do that seems unnecessary to me. A user can have a binding/command with glue to preface a directory if they want, so that they could use `:sesh work` to write to `~/.config/luakit/sessions/work`. Adding a function (allbeit a small one) to a core module so that a user can avoid doing a string concatenation seems misguided to me.



> > Reusing tabs (webviews) which are already loaded would probably be faster.

My point here was that if you're wanting to trade the current state for a different one, 
rather than throwing away everything that you have open, and then loading everything that you want from scratch,
(i imagine) it'd be less resource intensive (in terms of starting up new webview processes, etc) to figure out what resources you have loaded, and what resources you need, and retask the already-loaded stuff where possible. It would _certainly_ be the biggest project of those suggested so far, and not necessarily worth the effort..

> and tab groups work.

There is a module `tabgroups` which is not loaded by default (you'll have to uncomment it in your `rc.lua`),
which gives you a menu of "tab groups". One of these can be loaded into the current window at a given time, while the rest just hang out in the background. If you're familiar with tmux, a tabgroup is roughly analogous to a tmux session, while a tab is roughly analogous to a tmux window.

> If you feel like the above suggestions are "better" or can be imagined in the same way as hacking on the session module I'll take your suggestion.

Well, i still don't have a firm understanding of how you want to work, so i'm not at all sure that any of this is good advice.
I feel the tabgroups module is worth looking at, as it would be quick to do and might give you the functionality you're after.
Reusing webviews would probably be a lot of work, but would offer the most control and be the lightest option in terms of system resources.

**Comment by ldelossa** [comment]
@taobert sorry i meant to add that your assumption about my use case was indeed correct. 

After working all day, I (and I assume others) load up their browsers with tabs. It would be nice to blast them all away, and start with your "default" set based on your current environment such as "work". 

Its also very nice just a new os boot. I often dont want to restore all my tabs on a restart, i just want to restore my work tabs (email, calendar, gh notifications, etc...) 

Im definitely seeing your point around simply exporting the the "restore_file" method and leaning on user defined bindings, making the core module just whats needed from a functionality point of view and letting the users use it as they see fit.
